**SyntaxError**
- Occurs when Python encounters incorrect syntax
- Usually due to typos or not knowing Python well enough

**NameError**
- This occurs when a variable is not defined, i.e. it hasn't been assigned

**TypeError**<br>
Occurs when:
1. An operation or function is applied to the wring type
2. Python cannot interpret an operation on two data types

**IndexError**<br>
Occurs when you try to access an element in a list using an invalid index

**ValueError**<br>
This occurs when a built-in operation or function receives an argument that has the right type but an inappropriate value

**KeyError**<br>
This occurs when a dictionary does not have a specific key.

**AttributeError**<br>
This occurs when a variable does not have an attribute

**Raise your own exception**<br>
In python we can also throw errors using the raise keyword. this is helpful when creating your own kinds of exception and error message
```Python
raise ValueError('invalid value') #equivalent to throw in Java
```
```Python
def colorize(text, color):
    colors = ("cyan", "yellow", "blue")
    if type(color) is not str:
        raise TypeError("color must be instance of str")
    if type(text) is not str:
        raise TypeError("text must be instance of str")
    if color not in colors:
        raise ValueError("color is invalid color")
    print(f"Printed {text } in {color}")
```
**Handle Errors: try/except blocks**
```Python
d = {"name": "Ricky"}

def get(d, key):
    try:
        return d[key]
    except KeyError:
        return None

try:
    num = int(input('please enter a number:'))
except ValueError as err:
    print("That's not a number!")
    print(err)
else:
    print("I'm in the else")
finally:
    print("Runs no matter what")

def divide(a,b):
    try:
        result = a/b
    except (ZeroDivisionError, TypeError) as err:
        print("Something went wrong")
        print(err)
    else:
        print(f"{a} divided by {b} is {result}")
```
**Debugging with Pdb(Python Debugger)**<br>
To set breakpoints in our code we can use pdb by inserting this line:
```Python
import pdb; pdb.set_trace() #commonly on one line
```
```Python
import pdb
first = "First"
second = "Second"
pdb.set_trace()
result = first +second
third = "Third"
result += third
print(result)
```
```Python
#common PDB commands
# l (list)
# n (next_line)
# p (print), if PDB commands is conflict with the variable name, use p plus variable name. e,g. p l
# c (continue - finishes debugging)
```
